File: source/array/python/001_TwoSum.py
# -*- coding: utf-8 -*-


class Solution:
    def twoSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """

        rels = {}
        for pos in range(len(nums)):
            if nums[pos] in rels:
                return [rels[nums[pos]], pos]
            else:
                rels[target-nums[pos]] = pos

        return []

s = Solution()
print(s.twoSum([2, 7, 11, 15], 9))

File: source/array/c++_2/001_Two_Sum.cpp
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        vector<int> ret;
        unordered_map<int, int> v;
        
        for (unsigned i = 0; i < nums.size(); i++)
            v.insert({nums[i], i});  // <key:element, value:index>
        
        for (auto it = nums.begin(); it != nums.end(); ++it) {
            auto it2 = v.find(target - *it);
            // the index in the vector can not be the same as in unordered_map
            if (it2 != v.end() && (it - nums.begin()) != it2->second) {
                ret.push_back(it - nums.begin());
                ret.push_back(it2->second);
                break;
            }
        }
        return ret;
    }
};

